File: scripts/orderBookAnalyzer.ts
import { Decimal } from "decimal.js";
import { BINANCE_CONFIG } from "../configs/Binance_config.js";
import ExchangeClient from "../exchange/ExchangeClient.js";
import OrderBookAnalyzer from "../exchange/OrderBookAnalyzer.js";

function parseArg(flag: string): string | undefined {
  const arg = process.argv.find((a) => a.startsWith(flag + "="));
  return arg?.split("=")[1];
}
async function main() {
  const symbol = process.argv[2];
  const depth = Number(parseArg("--depth"));

  const client = await ExchangeClient.fromConfig(BINANCE_CONFIG);
  const orderBook = await client.fetchOrderBook(symbol, depth);
  if (!orderBook) return;
  const analyzer = new OrderBookAnalyzer(orderBook);

  console.log("=".repeat(65));
  console.log(symbol + " Order Book Analyzer");
  const today = new Date();
  console.log(
    "Timestamp: " +
      (!isNaN(orderBook.timestamp) ? orderBook.timestamp : today.toISOString()),
  );
  console.log("=".repeat(65));
  const asset = symbol.split("/")[0];
  console.log(
    "Best Bid: $" +
      orderBook.bestBid[0] +
      " x " +
      orderBook.bestBid[1] +
      " " +
      asset,
  );
  console.log(
    "Best Ask: $" +
      orderBook.bestAsk[0] +
      " x " +
      orderBook.bestAsk[1] +
      " " +
      asset,
  );
  console.log("Mid Price: $" + orderBook.midPrice);
  console.log(
    "Spread:    $" + orderBook.spread + ` (${orderBook.spreadBps}bps)`,
  );

  console.log("=".repeat(65));

  console.log("Depth (within 10 bps):");
  const bids = analyzer.depthAtBps("bid", 10);
  const bidsCost = analyzer.walkTheBook("sell", bids);
  console.log(`    Bids: ${bids} ${asset} ($${bidsCost.totalCost})`);
  const asks = analyzer.depthAtBps("bid", 10);
  const asksCost = analyzer.walkTheBook("buy", bids);
  console.log(`    Asks: ${asks}  ${asset} ($${asksCost.totalCost})`);
  const imbalance = analyzer.imbalance(10);
  console.log(
    `Imbalance: ${imbalance} (${imbalance === 0 ? "no" : "slight " + (imbalance > 0 ? "bid" : "ask")} pressure)`,
  );

  console.log("=".repeat(65));

  console.log("BUY 2");
  const buy2 = analyzer.walkTheBook("buy", new Decimal(2));

  console.log(`Avg price:  $${buy2.avgPrice}`);
  console.log(`Slippage:   ${buy2.slippageBps} bps`);
  console.log(`Levels:     ${buy2.levelsConsumed}`);

  console.log("BUY 10");
  const buy10 = analyzer.walkTheBook("buy", new Decimal(10));

  console.log(`Avg price:  $${buy10.avgPrice}`);
  console.log(`Slippage:   ${buy10.slippageBps} bps`);
  console.log(`Levels:     ${buy10.levelsConsumed}`);

  console.log("=".repeat(65));

  console.log(
    `Effective spread (2): ${analyzer.effectiveSpread(new Decimal(2))}bps`,
  );
}
main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: scripts/order_book_analyzer.py
# scripts/analyze_order_book.py

import sys
from decimal import Decimal
from datetime import datetime

from configs.config import BINANCE_CONFIG
from exchange.exchange_client import ExchangeClient
from exchange.order_book_analyzer import OrderBookAnalyzer


def parse_arg(flag: str) -> str | None:
    prefix = flag + "="
    arg = next((a for a in sys.argv if a.startswith(prefix)), None)
    return arg.split("=")[1] if arg else None


def main():
    if len(sys.argv) < 3:
        print("Usage: analyze_order_book.py <SYMBOL> [--depth=N]")
        sys.exit(1)

    symbol = sys.argv[1]
    depth = int(parse_arg("--depth") or 20)

    client = ExchangeClient(BINANCE_CONFIG)
    order_book = client.fetch_order_book_rest(symbol, depth)
    if not order_book:
        return

    analyzer = OrderBookAnalyzer(order_book)
    asset = symbol.split("/")[0]
    timestamp = (
        order_book.timestamp
        if order_book.timestamp
        else datetime.now().isoformat()
    )

    print("=" * 65)
    print(f"{symbol} Order Book Analyzer")
    print(f"Timestamp: {timestamp}")
    print("=" * 65)

    print(
        f"Best Bid: ${order_book.best_bid[0]} x {order_book.best_bid[1]} {asset}")
    print(
        f"Best Ask: ${order_book.best_ask[0]} x {order_book.best_ask[1]} {asset}")
    print(f"Mid Price: ${order_book.mid_price}")
    print(f"Spread:    ${order_book.spread} ({order_book.spread_bps}bps)")

    print("=" * 65)

    print("Depth (within 10 bps):")
    bids = analyzer.depth_at_bps("bid", 10)
    bids_cost = analyzer.walk_the_book("sell", bids)
    print(f"    Bids: {bids} {asset} (${bids_cost['total_cost']})")

    asks = analyzer.depth_at_bps("ask", 10)       # ← bug fix: "ask" not "bid"
    asks_cost = analyzer.walk_the_book(
        "buy", asks)    # ← bug fix: asks not bids
    print(f"    Asks: {asks} {asset} (${asks_cost['total_cost']})")

    imbalance = analyzer.imbalance(10)
    if imbalance == 0:
        pressure = "no"
    elif imbalance > 0:
        pressure = "bid"
    else:
        pressure = "ask"
    print(f"Imbalance: {imbalance} ({pressure} pressure)")

    print("=" * 65)

    print("BUY 2")
    buy2 = analyzer.walk_the_book("buy", Decimal(2))
    print(f"Avg price:  ${buy2['avg_price']}")
    print(f"Slippage:   {buy2['slippage_bps']} bps")
    print(f"Levels:     {buy2['levels_consumed']}")

    print("BUY 10")
    buy10 = analyzer.walk_the_book("buy", Decimal(10))
    print(f"Avg price:  ${buy10['avg_price']}")
    print(f"Slippage:   {buy10['slippage_bps']} bps")
    print(f"Levels:     {buy10['levels_consumed']}")

    print("=" * 65)
    print(f"Effective spread (2): {analyzer.effective_spread(Decimal(2))}bps")


if __name__ == "__main__":
    try:
        main()
    except Exception as err:
        print(err, file=sys.stderr)
        sys.exit(1)
